sdk/src/azure/platform/az_pahoasync.c

```diff
@@ -0,0 +1,619 @@
+
+  az_result ret = AZ_OK;
+  az_mqtt5* me = (az_mqtt5*)mqtt5;
+  MQTTAsync_createOptions create_opts = MQTTAsync_createOptions_initializer5;
+  MQTTAsync_connectOptions conn_opts = MQTTAsync_connectOptions_initializer5;
+  MQTTAsync_SSLOptions ssl_opts = MQTTAsync_SSLOptions_initializer;
+
+  if (me->_internal.options.disable_tls == 0)
+  {
+    ssl_opts.verify = 1;
+    bool use_os_certs
+        = (az_span_ptr(me->_internal.options.certificate_authority_trusted_roots) == NULL);
+
+    if (!use_os_certs)
+    {
+      ssl_opts.disableDefaultTrustStore = 1;
+      ssl_opts.CApath
+          = (const char*)az_span_ptr(me->_internal.options.certificate_authority_trusted_roots);
+    }
+
+    ssl_opts.keyStore = (const char*)az_span_ptr(connect_data->certificate.cert);
+    ssl_opts.privateKey = (const char*)az_span_ptr(connect_data->certificate.key);
+  }
+
+  if (az_span_ptr(connect_data->username) != NULL)
+  {
+    conn_opts.username = (const char*)az_span_ptr(connect_data->username);
+    conn_opts.password = (const char*)az_span_ptr(connect_data->password);
+  }
+
+  char port_buffer[SERVER_URI_PORT_MAX_LENGTH];
+  sprintf(port_buffer, "%d", connect_data->port);
+  az_span port_span = az_span_create_from_str(port_buffer);
+
+  int32_t server_uri_length
+      = az_span_size(connect_data->host) + SERVER_URI_COLON_LENGTH + SERVER_URI_PORT_MAX_LENGTH + 1;
+  char server_uri_buffer[server_uri_length];
+  az_span server_uri = az_span_create((uint8_t*)server_uri_buffer, server_uri_length);
+  az_span server_uri_temp = server_uri;
+
+  server_uri_temp = az_span_copy(server_uri_temp, connect_data->host);
+  server_uri_temp = az_span_copy_u8(server_uri_temp, ':');
+  server_uri_temp = az_span_copy(server_uri_temp, port_span);
+  server_uri_temp = az_span_copy_u8(server_uri_temp, '\0');
+
+  _az_RETURN_IF_FAILED(_az_result_from_pahoasync(MQTTAsync_createWithOptions(
+      me->_internal.pahoasync_handle,
+      (const char*)az_span_ptr(server_uri),
+      (const char*)az_span_ptr(connect_data->client_id),
+      MQTTCLIENT_PERSISTENCE_NONE,
+      NULL,
+      &create_opts)));
+
+  if (me->_internal.pahoasync_handle == NULL)
+  {
+    ret = AZ_ERROR_OUT_OF_MEMORY;
+    return ret;
+  }
+
+  _az_RETURN_IF_FAILED(_az_result_from_pahoasync(MQTTAsync_setCallbacks(
+      *me->_internal.pahoasync_handle,
+      mqtt5,
+      _az_pahoasync_on_disconnect,
+      _az_pahoasync_on_message,
+      NULL)));
+
+  conn_opts.keepAliveInterval = AZ_MQTT5_DEFAULT_MQTT_CONNECT_KEEPALIVE_SECONDS;
+  conn_opts.cleansession = false;
+  conn_opts.onSuccess5 = _az_pahoasync_on_connect_success;
+  conn_opts.onFailure5 = _az_pahoasync_on_connect_failure;
+  conn_opts.context = me;
+  conn_opts.ssl = &ssl_opts;
+
+  _az_RETURN_IF_FAILED(
+      _az_result_from_pahoasync(MQTTAsync_connect(*me->_internal.pahoasync_handle, &conn_opts)));
+
+  return ret;
+}
+
+AZ_NODISCARD az_result az_mqtt5_outbound_sub(az_mqtt5* mqtt5, az_mqtt5_sub_data* sub_data)
```

Comment:
Will we still get the mid filled out in the sub_data once this function returns?

Reply:
(same question for pub)

Reply:
Thanks for pointing that out! I've added the changes to fill out the message id